jetbrains-core/src/software/aws/toolkits/jetbrains/core/gettingstarted/SetupAuthenticationDialog.kt

```diff
@@ -0,0 +1,349 @@
+                        init()
+                    }
+
+                    override fun createCenterPanel() = idcRolePopup(tokenProvider)
+                }.show()
+            }
+
+            SetupAuthenticationTabs.BUILDER_ID -> {
+                loginSso(project, SONO_URL, SONO_REGION, scopes)
+            }
+
+            SetupAuthenticationTabs.IAM_LONG_LIVED -> {
+            }
+        }
+
+        close(OK_EXIT_CODE)
+    }
+
+    private fun selectedTab() = wrappers.entries.firstOrNull { (_, wrapper) -> wrapper == rootTabPane.selectedComponent }?.key
+        ?: error("Could not determine selected tab")
+
+    private fun idcTab() = panel {
+        row(message("gettingstarted.setup.idc.startUrl")) {
+            textField()
+                .comment(message("gettingstarted.setup.idc.startUrl.comment"))
+                .errorOnApply(message("gettingstarted.setup.error.not_empty")) { it.text.isBlank() }
+                .bindText(state.idcTabState::startUrl)
+        }
+
+        row(message("gettingstarted.setup.idc.region")) {
+            comboBox(
+                AwsRegionProvider.getInstance().let { provider ->
+                    provider.allRegions().values.filter { provider.isServiceSupported(it, "sso") }
+                },
+                SimpleListCellRenderer.create("null") { it.displayName }
+            ).bindItem(state.idcTabState::region.toNullableProperty())
+                .errorOnApply(message("gettingstarted.setup.error.not_selected")) { it.selected() == null }
+        }
+    }
+
+    private fun idcRolePopup(tokenProvider: SdkTokenProvider) = panel {
+        row {
+            label(message("gettingstarted.setup.idc.roleLabel"))
+        }
+
+        row {
+            val combo = AsyncComboBox<RoleInfo> { label, value, _ ->
+                value ?: return@AsyncComboBox
+                label.text = "${value.roleName()} (${value.accountId()})"
+            }
+            Disposer.register(myDisposable, combo)
+            combo.proposeModelUpdate { model ->
+                val token = tokenProvider.resolveToken().token()
+                val client = AwsClientManager.getInstance().createUnmanagedClient<SsoClient>(
+                    AnonymousCredentialsProvider.create(),
+                    Region.of(state.idcTabState.region.id)
+                )
+
+                client.listAccounts { it.accessToken(token) }
+                    .accountList()
+                    .flatMap { account ->
+                        client.listAccountRoles {
+                            it.accessToken(token)
+                            it.accountId(account.accountId())
+                        }.roleList()
+                    }.forEach {
+                        model.addElement(it)
+                    }
+            }
+
+            cell(combo)
+                .bindItem(state.idcTabState::roleInfo)
+        }
+    }
+
+    private fun builderIdTab() = panel {
```

Comment:
Is there a way to have a common function for these panels?

Reply:
what would be the common part?

Reply:
Nvm i got confused with the other tabs.